src/core/Akka/Dispatch/AbstractDispatcher.cs

```diff
@@ -241,7 +441,62 @@ public virtual void SystemDispatch(ActorCell cell, Envelope envelope)
         /// <param name="cell">The ActorCell belonging to the actor who's attaching to this dispatcher.</param>
         public virtual void Attach(ActorCell cell)
         {
-            
+            Register(cell);
+            RegisterForExecution(cell.Mailbox, false, true);
+        }
+
+        /// <summary>
+        /// INTERNAL API 
+        /// 
+        /// If you override it, you must still call the base method. But only ever once. See <see cref="Attach"/> for only invocation.
+        /// </summary>
+        /// <param name="actor">The actor we're registering</param>
+        internal virtual void Register(ActorCell actor)
+        {
+            if (DebugDispatcher) Actors.Value.Put(this, (IInternalActorRef)actor.Self);
+            AddInhabitants(1);
+        }
+
+        /// <summary>
+        /// IRunnable used for running the mailbox
+        /// </summary>
+        private sealed class MailboxRunAction : IRunnable // TODO: pool these
+        {
+            private readonly Mailbox _mbox;
+
+            public MailboxRunAction(Mailbox mbox)
+            {
+                _mbox = mbox;
+            }
+
+            public void Run()
+            {
+                _mbox.Run();
+            }
+        }
+
+        /// <summary>
+        /// INTERNAL API
+        /// 
+        /// Suggest to register the provided mailbox for execution
+        /// </summary>
+        /// <param name="mbox">The mailbox</param>
+        /// <param name="hasMessageHint">Do we have any messages?</param>
+        /// <param name="hasSystemMessageHint">Do we have any system messages?</param>
+        /// <returns><c>true</c> if the <see cref="Mailbox"/> was scheduled for execution, otherwise <c>false</c>.</returns>
+        internal bool RegisterForExecution(Mailbox mbox, bool hasMessageHint, bool hasSystemMessageHint)
+        {
+            if (mbox.CanBeScheduledForExecution(hasMessageHint, hasSystemMessageHint)) //This needs to be here to ensure thread safety and no races
+            {
+                if (mbox.SetAsScheduled())
```

Comment:
Second check here - only start the run if we can successfully set the mailbox to `Scheduled`. Several reasons why this may not happen:
1. Mailbox has been closed
2. Mailbox is already scheduled